### PR TITLE
Scheduler: remove pkg/features dependency from NodeResources plugins

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is an alternative to https://github.com/kubernetes/kubernetes/pull/98545, which attempts to decouple featuregate dependencies from plugin logic in the scheduler.

This PR takes the approach of parsing featuregates in the registry initialization, vs the other PR which injects it in API conversion

**Which issue(s) this PR fixes**:

Ref #89930

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig scheduling

### CODE REVIEW

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: As we discussed in https://github.com/kubernetes/kubernetes/pull/98545#issuecomment-770056695, prefer to use a map or WithXYZOption pattern as the input params. Otherwise, once some features reach to GA, the signature of this function needs to be updated.

Comment (damemi) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: Yeah, I like the `WithXYZOption` approach, something like `.WithFeaturegates(map[featuregate.Feature]bool)`?

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: Are we going to have the featuregate package in staging?

What if we do: `WithEnableXYZ`, where xyz is the specific feature, so each plugin have an explicit flag that gets set.

Comment (damemi) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: @Huang-Wei actually pointed out that `featuregate.Feature` is from component-base already: https://github.com/kubernetes/component-base/blob/1722fee/featuregate/feature_gate.go#L33. But the feature name constants are still in `pkg/features` so I don't know if that would work.

Using `WithEnableXYZFeature` would probably be easier to work with

**Comment by damemi** [comment]
@Huang-Wei @ahg-g I pushed an update here with what we could accomplish updating the PluginFactory signature. Unfortunately, we can't really get rid of the registry wrapping doing that unless we want to update the framework.Plugin interface to call these updates when the factory funcs are initialized.

What I also did was open https://github.com/kubernetes/kubernetes/pull/98675, which adds featuregates to the framework handle. This requires a lot less changes and is less disruptive to other plugins, please take a look

**Comment by Huang-Wei** [comment]
Thanks @damemi. #98675 looks more promising. I added some comments there.

**Comment by ehashman** [comment]
Are we closing this in favour of https://github.com/kubernetes/kubernetes/pull/98675 ?

File: pkg/scheduler/framework/plugins/registry.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/registry.go:61]: If we are going to wrap anyways, we should keep the same signature for `PluginFactory`, as callers of the Registry are not going to pass any option.

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/registry.go:61]: +1

**Comment by ahg-g** [comment]
I am in favor of this approach, I think this is the simplest one with the least amount of plumbing especially if we drop the change to PluginFactor signature which is not really needed.

**Comment by damemi** [comment]
> I am in favor of this approach, I think this is the simplest one with the least amount of plumbing especially if we drop the change to PluginFactor signature which is not really needed.

I agree. Reviewing our options:
1. https://github.com/kubernetes/kubernetes/pull/98545 - injecting args into internal conversion
2. (this pr)
  a. [first commit](https://github.com/kubernetes/kubernetes/pull/98583) - simply wrap the plugin `New` functions
  b. [second commit](https://github.com/kubernetes/kubernetes/pull/98583/commits/76f2a2c118de239b546747b00bb7d0a60ce7d404) - updating PluginFactory
3. https://github.com/kubernetes/kubernetes/pull/98675 - add feature gate to framework handle

It sounds like we prefer `2a`, for being the least intrusive. I am fine with that, and if we have a consensus on it I'll go ahead and close the other PRs. This can then serve as an example for wiring featuregates into plugins in the future.

**Comment by Huang-Wei** [comment]
> It sounds like we prefer `2a`, for being the least intrusive. I am fine with that, and if we have a consensus on it I'll go ahead and close the other PRs.

SGTM.

**Comment by ehashman** [comment]
Thanks for the great alternatives writeup! I would be inclined to go with the least intrusive solution, for whatever it's worth.

**Comment by damemi** [comment]
Glad for the feedback, I removed the other 2 commits from this PR and it should be back to just updating the `New` function params and wrapping them in the registry. Please double check for me, thanks!

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: do we still want to use the With Pattern (without changing the PluginFactory signature) just so we avoid mixing the flags (values passed to a series of bool parameters can easily be switched mistakenly)

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:68]: +1
Make sure the `WithMyFeatureFooEnabled` functions belong to each plugin package.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: If we want to totally eliminate the dependency on `features`, we can just pass `true`. Or, it's tolerated to have this deps on tests.

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: Ideally you need to test both, true and false, and not depend on the feature gate value at all.

Comment (damemi) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: Prior to this, the tests already depended on the feature gate value anyway (since it was interpreted internally by the plugin)

Though, I think there are essentially tests for both cases (I noticed some specific tests around pod overhead that failed without the parameter, while most other tests didn't rely on it)

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: For this particular case, whether the feature gate is enabled or not doesn't matter, the semantics is like "hey, I'm a consumer of NodeFit, I just want to derive whatever featuregate is set in your side".

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: but Fit shouldn't know about feature gates. So this test should be either passing it from the feature gate or just pass whatever value you want.

I presume this test doesn't make use of PodOverhead, so it doesn't matter whether we enable the feature or not.

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:105]: I think we should completely cut the dependency of the plugins pkg on the features pkg. Where it makes sense, we should test both values, true and false. Perhaps in this specific case since it is not really testing the plugin in question it is ok to just select any.

**Comment by damemi** [comment]
Updated this to:
* remove `pkg/features` dep from the plugins unit test (I think this was actually left in on accident, thanks for the catch)
* Switch out boolean parameters for `With..` functions that are more descriptive. Tried to re-use where I could with the score plugins, but Fit needed its own

Any other suggestions please let me know!

File: pkg/scheduler/core/generic_scheduler_test.go
Comment (alculquicondor) [inline pkg/scheduler/core/generic_scheduler_test.go:1313]: this is a test, do we actually require the options?

File: pkg/scheduler/framework/plugins/noderesources/fit.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/fit.go:117]: should we have the beta features enabled by default? It might save some surprises in tests or other unforeseen users

File: pkg/scheduler/framework/plugins/noderesources/fit.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/fit.go:98]: Just to save the prefix `Fit`, what about having a generic struct that holds all the booleans and to which all options in this package apply to?

Then you copy over all the needed flags from that struct to the actual objects (or just make it an unnamed member). I know not all flags apply to all objects, but 🤷‍♂️

File: pkg/scheduler/framework/plugins/registry.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/registry.go:21]: just `feature` if there is no collision.

File: pkg/scheduler/scheduler_test.go
Comment (alculquicondor) [inline pkg/scheduler/scheduler_test.go:773]: let's not pass a feature if the test doesn't depend on it.

Site note: we should push for graduation of these features next release 😂

File: pkg/scheduler/core/generic_scheduler_test.go
Comment (ahg-g) [reply pkg/scheduler/core/generic_scheduler_test.go:1313]: I would test with options in the plugin unit tests only

File: pkg/scheduler/framework/plugins/noderesources/fit.go
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/noderesources/fit.go:117]: this is going to be hard to track though, and so will likely cause inconsistencies.

File: pkg/scheduler/framework/plugins/noderesources/fit.go
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/noderesources/fit.go:98]: yeah, lets have a single one per plugin directory, they will likely share most of the features.

File: pkg/kubelet/lifecycle/predicate.go
Comment (damemi) [reply pkg/kubelet/lifecycle/predicate.go:232]: Removing the features dep from the plugin means this is the only way to pass that value in to `Fits`, so we need to make the change with it

Comment (ahg-g) [reply pkg/kubelet/lifecycle/predicate.go:232]: right, I am just suggesting to do it in a follow up PR so we get sig-node to review a significantly smaller one.

Comment (ahg-g) [reply pkg/kubelet/lifecycle/predicate.go:232]: btw, I think this `Fit` function is only used by kubelet

Comment (damemi) [reply pkg/kubelet/lifecycle/predicate.go:232]: I understand, but my point is if we don't react to the change in this PR then any kubelet/lifecycle tests that rely on this feature gate will fail

Comment (damemi) [reply pkg/kubelet/lifecycle/predicate.go:232]: And not updating the `Fits` signature means we need to somehow pass a podOverhead value (hardcoded?) to `computePodResourceRequest`. That's the function shared by all the scorers. I think it is just reasonable to get sig-node to be aware of this change now

Comment (alculquicondor) [reply pkg/kubelet/lifecycle/predicate.go:232]: While unfortunate that we can't isolate kubelet changes, I agree with Mike.

Comment (ahg-g) [reply pkg/kubelet/lifecycle/predicate.go:232]: Can't we update Fits as follows:
```
// Fits checks if node have enough resources to host the pod.
func Fits(pod *v1.Pod, nodeInfo *framework.NodeInfo) []InsufficientResource {
	return fitsRequest(computePodResourceRequest(pod, utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead)), nodeInfo, nil, nil)
}
```

Comment (damemi) [reply pkg/kubelet/lifecycle/predicate.go:232]: @ahg-g that would keep the featuregate dep in the plugin package. We could do that now, and then actually remove the dep in a follow up if you'd like

Comment (ahg-g) [reply pkg/kubelet/lifecycle/predicate.go:232]: yes, removing the dependency is basically a change to the build file and removing the import. It is up to you, but this is my advise if you want to make faster progress :)

File: pkg/scheduler/framework/plugins/noderesources/fit.go
Comment (damemi) [reply pkg/scheduler/framework/plugins/noderesources/fit.go:98]: An easy way to do this would be to just embed a `resourceAllocationScorer` in the `Fit` struct. It would only be used for the featuregate flags, but then we don't need to add a new type just for these. wdyt?

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/noderesources/fit.go:98]: -1

Better have an isolated struct just for flags

**Comment by damemi** [comment]
Pushed a new commit that creates a `featureGates` struct that holds the featuregate values for the nodeResources plugins. The struct is embedded in `resourceAllocationScorer` and `Fit`, so that all the scoring plugins and the Fit plugin can share this type. Ptal!

File: pkg/kubelet/lifecycle/predicate.go
Comment (damemi) [reply pkg/kubelet/lifecycle/predicate.go]: Perhaps just splitting the change into a separate commit will ease the work for the kubelet reviewers, it is such a minor change that I think it will be quicker to get it done here if that's all right

File: pkg/scheduler/core/generic_scheduler_test.go
Comment (alculquicondor) [inline pkg/scheduler/core/generic_scheduler_test.go:1319]: put these in a single line

**Review by alculquicondor — COMMENTED** [review]
lgtm after nit

**Comment by alculquicondor** [comment]
good for squash

**Comment by alculquicondor** [comment]
/hold

I just realized: why are we doing the `WithOption` pattern? This is useful when we want to hide the changes in a function signature. However, in this case, we still need a wrapper around the constructors because the registry requires a specific signature.

Instead, we can do:

```
func NewPlugin(args runtime.Object, handle framework.Handle, features ResourceFeatures)
```

This is similar to a clientset method `Update(ctx, obj, metav1.UpdateOptions{})` and gives us the same "named" parameters that we wanted to avoid mixing the booleans, being simpler.

**Comment by ahg-g** [comment]
> /hold
> 
> I just realized: why are we doing the `WithOption` pattern? This is useful when we want to hide the changes in a function signature. However, in this case, we still need a wrapper around the constructors because the registry requires a specific signature.
> 
> Instead, we can do:
> 
> ```go
> func NewPlugin(args runtime.Object, handle framework.Handle, features ResourceFeatures)
> ```
> 
> This is similar to a clientset method `Update(ctx, obj, metav1.UpdateOptions{})` and gives us the same "named" parameters that we wanted to avoid mixing the booleans, while being simpler.

The other proposal was to bake that in the internal type of the args which triggered too many changes. The advantage was that the definition was localized and the framework interface was kept intact.

The difference between what we are trying to do here and `UpdateOptions` is that features graduate and so fields will continue to be added and removed from a type that will be part of the framework interface, which is supposed to be relatively stable (I am assuming `ResourceFeatures` will include fields like `PodOverheadEnabled`, which will be removed once the feature graduates). But may be that is ok since it is a featuregate by definition?

**Comment by damemi** [comment]
> The difference between what we are trying to do here and UpdateOptions is that features graduate and so fields will continue to be added and removed from a type that will be part of the framework interface, which is supposed to be relatively stable (I am assuming ResourceFeatures will include fields like PodOverheadEnabled, which will be removed once the feature graduates). But may be that is ok since it is a featuregate by definition?

I agree with Abdullah on this. Putting these fields into an external type means more API maintenance and less flexibility as feature gates come and go. And the changes necessary to update `PluginFactory` seemed excessive when we tried that approach.

From a usability standpoint, I don't really see a difference between the `WithOption` funcs and adding an extra param. The benefit to `WithOption` is that we can add an optional list (so not every caller has to react to the change) and still get the clarity of what's being set (as opposed to the list of bools).

**Comment by alculquicondor** [comment]
> And the changes necessary to update PluginFactory seemed excessive when we tried that approach.

I'm not suggesting you change the PluginFactory. We should keep the wrappers.

> The difference between what we are trying to do here and UpdateOptions is that features graduate and so fields will continue to be added and removed from a type that will be part of the framework interface, which is supposed to be relatively stable (I am assuming ResourceFeatures will include fields like PodOverheadEnabled, which will be removed once the feature graduates)

We have to remove the `WithOptions` too, which has the same burden.

**Comment by ahg-g** [comment]
I am starting to feel that this whole approach of plumbing in feature gates will cause problems long term. Sorry if you answered this before, but how feasible is it to move `kube_features.go` to staging?

**Comment by damemi** [comment]
I think moving `kube_features` is using a much bigger hammer for this problem than necessary. This is a design issue with these plugins, not the features package, and moving it would be more disruptive across core components as well as add  work to maintain accurate versioning across repos which I expect would see pushback.

As opposed to component-helpers (which holds functions with demonstrable external utility and cross-component consumption), other internal components like API server or controller manager have no use for externalizing feature gate definitions (that I know of). We are making the decision to externalize modular portions of the scheduler, so the burden lies with us in our design.

If feature definitions are eventually moved to staging, great, but I don't think that should influence this goal, which is essentially defining a pattern for wiring arbitrary plugin parameters. The pain points we are seeing now in deciding on a design will be much easier for future plugins once there is an accepted consensus.

That then goes further to your point,
> what if the framework itself needs to access a feature?

where the same pattern we decide on for plugins can be abstracted a layer down to the framework (this is why I prefer `WithXYZOption` functions, because the framework itself is already designed with those).

However, we can't externalize the framework without first externalizing the plugins, and in my opinion the core plugins should serve as design examples for out-of-tree plugins just how the core scheduler is itself an implementation of the framework. Being a core component, it could initialize the framework using the gating option pattern we decide.

Maybe I am way off base, and if we would like to open the broader architectural discussion of moving feature gates to staging (this has been briefly discussed before in https://github.com/kubernetes/kubernetes/issues/90136), then we could certainly use the scheduler and descheduler as examples for that. But, having examined the several options we have seen so far, that it is not critical for us and I believe we are capable of solving it on our own.

**Comment by ahg-g** [comment]
/lgtm

Thanks Mike, you make a good point.

**Comment by damemi** [comment]
Thanks Abdullah! I appreciate the open discussion to ensure we get this done right.

/cc @derekwaynecarr 
Derek, could you please approve for kubelet? The only change is to add a feature gate parameter to the function call in https://github.com/kubernetes/kubernetes/pull/98583/files#diff-1046a664dcd001985cf5a688c28959d4bafb964343b0d9faf3ed88e1a14c801bR232

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:47]: it looks like this was already imported above :)

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go:27]: since there is no chance of conflict with the other "feature" package (the point is not to import it), you don't need the rename/alias. You can do the same in all pkg/scheduler/framework but the registry.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:103]: does any test fail with the feature set to false? If not, maybe we can just leave everything empty to not give the impression that we cared. Although.... PodOverhead is already enabled by default.

**Review by alculquicondor — COMMENTED** [review]
Good to squash. Last comment is optional.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (damemi) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go]: Tested, and no other tests fail without it so I removed the PodOverhead parameter... but since the function still needs to be wrapped, I kept the declaration just for simplicity. Does that sound ok?

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go]: perfect!

**Comment by ahg-g** [comment]
Thanks Mike!

/lgtm
/approve

**Comment by damemi** [comment]
rebased
/hold cancel

**Comment by damemi** [comment]
Rebased this (didn't notice that it never merged, sorry).. will need another lgtm if tests pass

**Comment by SergeyKanzhelev** [comment]
/lgtm 

for node..

**Comment by damemi** [comment]
ping @derekwaynecarr could you please approve from kubelet?

**Review by SergeyKanzhelev — APPROVED** [review]

**Review by ehashman — COMMENTED** [review]
/retest
/lgtm

**Comment by damemi** [comment]
rebased
/cc @mrunalp 
for kubelet approval please

**Review by ehashman — COMMENTED** [review]
/retest
/lgtm

**Pull request merged**
